packages/atclient/tests/test_stringutils.c

```diff
@@ -0,0 +1,174 @@
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include "atclient/stringutils.h"
+#include "atclient/atlogger.h"
+
+#define TAG "test_stringutils"
+
+int main()
+{
+    int ret = 1;
+
+    const unsigned long outlen = 4096;
+    char *out = (char *) malloc(sizeof(char) * outlen);
+    memset(out, 0, sizeof(char) * outlen);
+    unsigned long outolen = 0;
+
+    const unsigned long tokenlen = 4096;
+    const unsigned long tokenslen = 16;
+    char **tokens = malloc(sizeof(char *) * tokenslen);
+    memset(tokens, 0, sizeof(char *) * tokenslen);
+    for(int i = 0; i < tokenslen; i++)
+    {
+        *(tokens + i) = malloc(sizeof(char) * tokenlen);
```

Comment:
This is a very expensive split operation, since we are allocating the full length of the string for each split... instead, we should do this allocation inside the loop of the split function.

Reply:
I mean, you can malloc (current size + (num splits - 1 bytes)) then copy each split, inserting a null terminator in between. Freeing the memory should be relatively easy, iterate over the array and free each pointer.

Reply:
Actually, since we discard the delimiter, we can simply replace the delimiter with a null terminator and call it a day. There will be allocated memory in between the null terminator and the start of the next element if the delimiter is more than a one character, but that's fine.

Reply:
`realloc` can also be expensive but maybe works fine in this context to keep it simple (same pointer, no delimiters...)